File: haskell/Menu.hs
module Menu where

import Licao
import Data.Char(toLower)
import Controller (licoes)
import Text.Read(readMaybe)
import System.Exit (exitSuccess)
import Util(limpaTela, leCaractere)
import System.Posix.Internals (o_RDONLY)
import Sprites (exibeProgresso, formataRanking, aplicaCorInstrucao)
import Desafio (iniciaDesafio, Desafio (UmMinuto, DoisMinutos, TresMinutos), getRanking)

imprimeMenu :: IO()
imprimeMenu = do
    limpaTela
    menu <- readFile "../arteTxt/menu.txt"
    putStrLn menu 
    opcaoSelecionada <- getLine
    opcoesMenuPrincipal (map toLower opcaoSelecionada)

opcoesMenuPrincipal :: String -> IO()
opcoesMenuPrincipal o
    | o == "l" = listaLicoes
    | o == "d" = listaDesafios
    | o == "t" = exibeTutorial
    | o == "s" = sai
    | otherwise = imprimeMenu

listaLicoes :: IO ()
listaLicoes = do
    limpaTela

    dadosLicoes <- getDadosLicoes
    let todasLicoes = licoes dadosLicoes

    putStrLn (exibeProgresso (contaLicoesConcluidas todasLicoes))
    putStrLn (exibeLicoesConcluida todasLicoes)

    licoes <- readFile "../arteTxt/licoes.txt"
    putStrLn licoes
    comandoUsuario <- getLine
    if comandoUsuario == ""
        then imprimeMenu
        else case readMaybe comandoUsuario of
            Just n | n >= 1 && n <= 15 -> exibeLicao n todasLicoes
            _ -> do
                listaLicoes

exibeLicao :: Int -> [Licao] -> IO ()
exibeLicao idLicao licoes = do
    limpaTela
    let licaoSelecionada = licoes !! (idLicao - 1)
    instrucaoLicao <- readFile (instrucao licaoSelecionada)
    putStrLn $ aplicaCorInstrucao instrucaoLicao
    opcao <- leCaractere
    if opcao == 'i' then do
        limpaTela
        iniciaLicao licaoSelecionada
        setStatusLicao (show idLicao)
        voltaMenuLicoes
     else if opcao == '\n' then 
        listaLicoes
     else do
        exibeLicao idLicao licoes
    
listaDesafios :: IO ()
listaDesafios = do
    limpaTela
    desafios <- readFile "../arteTxt/desafios.txt"
    putStrLn desafios
    opcao <- getLine 
    opcoesMenuDesafio (map toLower opcao)
    voltaMenuDesafios

opcoesMenuDesafio :: String-> IO ()
opcoesMenuDesafio o
    | o == "1" = iniciaDesafio UmMinuto
    | o == "2" = iniciaDesafio DoisMinutos
    | o == "3" = iniciaDesafio TresMinutos
    | o == "r" = exibeRanking
    | o == "" = imprimeMenu
    | otherwise = listaDesafios

exibeRanking :: IO ()
exibeRanking = do
    limpaTela
    getRanking
    voltaMenuDesafios

exibeTutorial :: IO ()
exibeTutorial = do
    limpaTela
    tutorial <- readFile "../arteTxt/tutorial.txt"
    putStrLn tutorial
    voltaMenuPrincipal

sai :: IO()
sai = do
    limpaTela
    sai <- readFile "../arteTxt/sair.txt"
    putStrLn $ aplicaCorInstrucao sai
    exitSuccess

voltaMenuPrincipal :: IO()
voltaMenuPrincipal = do
    _ <- getLine
    imprimeMenu

voltaMenuLicoes :: IO()
voltaMenuLicoes = do
    _ <- getLine
    listaLicoes

voltaMenuDesafios :: IO()
voltaMenuDesafios = do
    _ <- getLine
    listaDesafios

File: prolog/Menu.pl
:- module(_, [imprime_menu/0, exibe_licao/1, lista_licoes/0, lista_desafios/0]).

:- use_module('./Utils.pl').
:- use_module('./Controller.pl').
:- use_module('./Licao.pl').
:- use_module('./Sprites.pl').
:- use_module('./Desafio.pl').

imprime_menu :-
    limpar_tela,
    ler_arquivo('../arteTxt/menu.txt'),
    ler_entrada(Entrada),
    downcase_atom(Entrada, OpcaoMenu),
    opcoes_menu_principal(OpcaoMenu).

opcoes_menu_principal('l') :- lista_licoes.
opcoes_menu_principal('d') :- lista_desafios.
opcoes_menu_principal('t') :- exibe_tutorial.
opcoes_menu_principal('s') :- sair.
opcoes_menu_principal(_) :- imprime_menu.


lista_licoes :-
    limpar_tela,
    conta_licoes_concluidas(Cont),
    exibe_progresso(Cont, Progresso),
    writeln(Progresso),
    le_status(Dados),
    exibe_licoes_concluidas(Dados, LicoesConcluidas),
    writeln(""),
    writeln(LicoesConcluidas),
    ler_arquivo('../arteTxt/licoes.txt'),
    ler_entrada(Entrada),
    (  Entrada = "" -> imprime_menu; number_string(NumeroLicao, Entrada),
        (   NumeroLicao >= 1, NumeroLicao =< 15 -> exibe_licao(NumeroLicao))
        ; lista_licoes).

exibe_licao(NumeroLicao) :-
    limpar_tela,
    licao(NumeroLicao, _Exercicios, Arquivo),
    ler_instrucao(Arquivo),
    ler_entrada(Entrada),
    downcase_atom(Entrada, Opcao),
    ( Entrada = "" -> lista_licoes; Opcao = 'i' -> inicia_licao(NumeroLicao)
    ; exibe_licao(NumeroLicao)).
    
lista_desafios :-
    limpar_tela,
    ler_arquivo('../arteTxt/desafios.txt'),
    ler_entrada(Entrada),
    le_ranking(Dados),
    ( Entrada = "" -> imprime_menu; Entrada = "r" -> 
        limpar_tela,
        exibe_ranking(Dados, Ranking),
        writeln(Ranking), ler_entrada(_), lista_desafios;
    number_string(NumeroDesafio, Entrada), opcoes_menu_desafios(NumeroDesafio)).

opcoes_menu_desafios(1) :- inicia_desafio(um_minuto).
opcoes_menu_desafios(2) :- inicia_desafio(dois_minutos).
opcoes_menu_desafios(3) :- inicia_desafio(tres_minutos).
opcoes_menu_desafios(_) :- lista_desafios.

exibe_tutorial:-
    limpar_tela,
    ler_arquivo('../arteTxt/tutorial.txt'),
    ler_entrada(_),
    imprime_menu.

sair :-
    limpar_tela,
    ler_arquivo('../arteTxt/sair.txt'),
    halt.